plugins/outputs/pulsar/pulsar.go

```diff
@@ -0,0 +1,378 @@
+			clientOptions.OperationTimeout = 30 * time.Second
+		} else {
+			clientOptions.OperationTimeout = operationTimeout
+		}
+
+	}
+
+	if p.AuthProvider == "tls" {
+		if p.TLSCertificatePath != "" && p.TLSPrivateKeyPath != "" {
+			clientOptions.Authentication = pulsar.NewAuthenticationTLS(p.TLSCertificatePath, p.TLSPrivateKeyPath)
+		} else {
+			return fmt.Errorf("tls auth specified but tls_certificate_path or tls_private_key_path is miising")
+		}
+
+	} else if p.AuthProvider == "token" {
+		if p.AuthToken != "" {
+			clientOptions.Authentication = pulsar.NewAuthenticationToken(p.AuthToken)
+		} else {
+			return fmt.Errorf("auth provider is token but auth_token is empty")
+		}
+
+	}
+
+	client, err := pulsar.NewClient(clientOptions)
+	if err != nil {
+		return err
+	}
+	p.client = client
+	return nil
+}
+
+func (p *Pulsar) Close() error {
+	if p.client != nil {
+		p.client.Close()
+	}
+	return nil
+}
+
+func (p *Pulsar) SampleConfig() string {
+	return sampleConfig
+}
+
+func (p *Pulsar) Description() string {
+	return "Configuration to send metrics to pulsar broker"
+}
+
+func (p *Pulsar) routingKey(metric telegraf.Metric) string {
+	if p.RoutingTag != "" {
+		key, ok := metric.GetTag(p.RoutingTag)
+		if ok {
+			return key
+		}
+	}
+
+	if p.RoutingKey == "measurement" {
+		return metric.Name()
+	}
+
+	if p.RoutingKey == "random" {
+		u, err := uuid.NewV4()
+		if err != nil {
+			p.Log.Errorf("Unable to generate uudid %s", err.Error())
+			return ""
+		}
+		return u.String()
+	}
+	return p.RoutingKey
+}
+func sendAsyncCallback(id pulsar.MessageID, message *pulsar.ProducerMessage, err error) {
+	if err != nil {
+		log.Printf("Failed to publish metrics : %s", err.Error())
+	}
+
+}
+
+func (p *Pulsar) Write(metrics []telegraf.Metric) error {
+	var batchSize uint
+	// Setting the batch size to metric_batch_size is unset
+	if p.BatchingMaxMessages == 0 {
+		batchSize = uint(len(metrics))
+	} else {
+		batchSize = p.BatchingMaxMessages
+	}
+	for _, metric := range metrics {
+		metric, topic := p.GetTopicName(metric)
+		producer, err := p.GetProducer(topic, batchSize)
+		if err == nil {
+			buf, err := p.serializer.Serialize(metric)
+			if err != nil {
+				p.Log.Debugf("Could not serialize metric: %v", err)
+				continue
+			}
+			m := pulsar.ProducerMessage{
+				Payload:   buf,
+				EventTime: metric.Time(),
+				Key:       p.routingKey(metric),
+			}
+
+			// Negative timestamps are not allowed by the Pulsar protocol.
+			if !metric.Time().Before(zeroTime) {
+				m.EventTime = metric.Time()
+			}
+			producer.SendAsync(context.Background(), &m, sendAsyncCallback)
+		} else {
+			p.Log.Errorf("Unable to create producer for topic %s : %s", topic, err.Error())
+			return err
+		}
+
+	}
+	// We flush and make sure all the metrics are published and acknowledged by the broker
+	// so that in case of failure the metrics are retried
+	for topic, producer := range p.producerCache {
```

Comment:
Wouldn't it be easier, more efficient and more safe to collect all producers used in the loop above and flush those instead of this indirect way?
Furthermore, would it make sense to flush all producers in parallel?